### PR TITLE
transport: fail NewStream() with better error message when conn is closed

### PR DESCRIPTION
**Pull request by menghanl**

### CODE REVIEW

**Comment by menghanl** [comment]
Found when debugging Test/KeepaliveClientStaysHealthyWithResponsiveServer #4424

**Comment by dfawley** [comment]
The test you changed now appears to be flaky.

File: internal/transport/http2_client.go [hunk 1]
```diff
@@ -743,6 +744,12 @@ func (t *http2Client) NewStream(ctx context.Context, callHdr *CallHdr) (_ *Strea
 		case <-t.goAway:
 			return nil, errStreamDrain
 		case <-t.ctx.Done():
+			t.mu.Lock()
+			err := t.closeErr
+			t.mu.Unlock()
+			if err != nil {
+				return nil, err
+			}
```
Comment (dfawley) [inline internal/transport/http2_client.go:752]: I don't believe the lock is necessary, but it's harmless in any case.  `t.cancel()` is called only after `t.closeErr` is set, and `t.closeErr` can only be set once, before `t.cancel()`.

File: internal/transport/transport_test.go [hunk 2]
```diff
@@ -780,10 +780,12 @@ func (s) TestGracefulClose(t *testing.T) {
 		go func() {
 			defer wg.Done()
 			str, err := ct.NewStream(ctx, &CallHdr{})
-			if err == ErrConnClosing {
-				return
-			} else if err != nil {
-				t.Errorf("_.NewStream(_, _) = _, %v, want _, %v", err, ErrConnClosing)
+			if err != nil {
+				select {
+				case <-ct.ctxDone:
```
Comment (dfawley) [inline internal/transport/transport_test.go:785]: This needs a comment to explain what it's needed for.

File: internal/transport/http2_client.go (hunk 1 above)
Comment (menghanl) [reply internal/transport/http2_client.go:752]: `t.ctx` is derived from a parent context, so `t.cancel()` isn't the only way to cancel `t.ctx`. And the read and write may race in the other case.
(That parent context is `ClientConn.ctx`, so this only happens when the ClientConn is closed)

File: internal/transport/transport_test.go (hunk 2 above)
Comment (menghanl) [reply internal/transport/transport_test.go:791]: Done

File: internal/transport/transport_test.go [hunk 3]
```diff
@@ -780,10 +780,18 @@ func (s) TestGracefulClose(t *testing.T) {
...
 			str, err := ct.NewStream(ctx, &CallHdr{})
-			if err == ErrConnClosing {
-				return
-			} else if err != nil {
-				t.Errorf("_.NewStream(_, _) = _, %v, want _, %v", err, ErrConnClosing)
+			if err != nil {
+				// The NewStream calls race with closing of the transport (after
+				// the last stream is closed).
+				// - if it is after the transport is closed (case <-ct.ctxDone),
+				//   we don't care about the error.
```
Comment (dfawley) [inline internal/transport/transport_test.go:787]: Why would the transport be closed? It was _gracefully_ closed, but we have an active stream (from line 755) so it should never close, should it?  Unless the server ends the stream, but it looks like it does not, until it receives the client's end-stream.

Comment (menghanl) [reply internal/transport/transport_test.go:787]: We start 200 goroutines to run this.
Some of them could be so slow that they run after the end of the line 755 stream.

File: internal/transport/http2_client.go (hunk 1 above)
Comment (menghanl) [reply internal/transport/http2_client.go:752]: How is this different?

And I don't like returning nil here. Even if it should never happen.

**Comment by dfawley** [comment]
Tests are failing: 

```
Error: ../internal/transport/http2_client.go:406:22: not enough arguments in call to t.controlBuf.finish
```

**Comment by menghanl** [comment]
This PR no longer works after moving `controlbuf.finish()` (#4447).

`NewStream()` will fail due to `controlbuf` error, before the transport is closed (it's a race), so it won't be able to return the transport error.

This will need more thoughts, and probably more changes.

**Comment by dfawley** [comment]
Can you file an issue to follow-up on this later?  (Unless you have something you're actively working on already.)

**Pull request closed without merging**